pkg/controller/podpreset/controller.go

```diff
@@ -0,0 +1,241 @@
+
+	c.podInformer = podInformer
+	c.podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addPod,
+			UpdateFunc: c.updatePod,
+		},
+	)
+	c.podLister = corelisters.NewPodLister(c.podInformer.GetIndexer())
+	c.podListerSynced = c.podInformer.HasSynced
+	c.initFn = c.initPod
+
+	return c, nil
+}
+
+func (c *Controller) addPod(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	glog.V(5).Infof("new Pod: %s with meta: %+v received", pod.GetName(), pod.ObjectMeta)
+	if needsInitialization(pod) {
+		glog.Infof("found an uninitialized pod: %+v", pod.Name)
+		if key, err := cache.MetaNamespaceKeyFunc(obj); err == nil {
+			c.podQueue.Add(key)
+		}
+	}
+}
+
+func (c *Controller) updatePod(old, new interface{}) {
+	pod := new.(*v1.Pod)
+	glog.V(5).Infof("Pod: %s with meta: %+v update received", pod.GetName(), pod.ObjectMeta)
+	if needsInitialization(pod) {
+		glog.Infof("found an existing uninitialized pod: %s", pod.GetName())
+		if key, err := cache.MetaNamespaceKeyFunc(new); err == nil {
+			c.podQueue.Add(key)
+		}
+	}
+}
+
+// Run runs the controller until the given stop channel can be read from.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer func() {
+		c.podQueue.ShutDown()
+	}()
+	glog.Info("Starting podpreset initializer")
+	defer glog.Infof("Shutting down podpreset initializer")
+
+	go c.podInformer.Run(stopCh)
+
+	// Wait for all caches to be synced, before processing items from the queue is started
+	if !cache.WaitForCacheSync(stopCh, c.podListerSynced) {
+		glog.Error(fmt.Errorf("Timed out waiting for pod cache to sync"))
+		return
+	}
+
+	// Wait for all caches to be synced, before processing items from the queue is started
+	if !cache.WaitForCacheSync(stopCh, c.podpresetListerSynced) {
+		glog.Error(fmt.Errorf("Timed out waiting for pod cache to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	// wait unitl we are told to stop
+	<-stopCh
+}
+
+func (c *Controller) runWorker() {
+	for c.processNext() {
+	}
+}
+
+func (c *Controller) processNext() bool {
+	// Wait until there is a new item in the working queue
+	key, quit := c.podQueue.Get()
+	if quit {
+		return false
+	}
+	// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+	// This allows safe parallel processing because two pods with the same key are never processed in
+	// parallel.
+	defer c.podQueue.Done(key)
+	// Initialize the incoming Pod
+	err := c.initFn(key.(string))
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) initPod(key string) error {
+	glog.V(5).Infof("got key: %v", key)
+
+	ns, podName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to parse key: %v :%v", key, err)
+	}
+
+	pod, err := c.podLister.Pods(ns).Get(podName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve pod %s in ns: %v : %v", podName, ns, err)
+	}
+
+	if !needsInitialization(pod) {
```

Comment:
It seems weird to @MHBauer and I that we need this check. It seems to me that if we get here and the pod no longer needs initialization something has Gone Wrong because we check if the pod still needs initialization in both places. Is there something we're not seeing?

Reply:
I believe you're correct, will remove.